pkg/binding/controller.go

```diff
@@ -477,67 +458,41 @@ func (c *Controller) processNextWorkItem(ctx context.Context) bool {
 	return true
 }
 
-func (c *Controller) reconcile(ctx context.Context, key util.Key) error {
+func (c *Controller) reconcile(ctx context.Context, objIdentifier util.ObjectIdentifier) error {
 	logger := klog.FromContext(ctx)
-	var obj runtime.Object
-	var err error
-	// special handling for binding resource as it is the only
-	// resource that is queued directly as key, without necessarily first
-	// existing as an object.
-	if util.KeyIsForBinding(key) {
-		return c.syncBinding(ctx, key) // this function logs through all its exits
-	}
-
-	if key.DeletedObject == nil {
-		obj, err = c.getObjectFromKey(key)
-		if err != nil {
-			// The resource no longer exist, which means it has been deleted.
-			if errors.IsNotFound(err) {
-				logger.Info("Object referenced from work queue no longer exists",
-					"object-name", key.NamespacedName, "object-gvk", key.GvkKey())
-				return nil
-			}
-			return err
-		}
-	} else {
-		obj = *key.DeletedObject
-	}
 
 	// special handling for selected API resources
 	// note that object is *unstructured.Unstructured so we cannot
 	// just use "switch obj.(type)"
-	if util.IsBindingPolicy(obj) {
-		if err := c.handleBindingPolicy(ctx, obj); err != nil {
+	if util.ObjIdentifierIsForBinding(objIdentifier) {
+		return c.syncBinding(ctx, objIdentifier) // this function logs through all its exits
+	} else if util.ObjIdentifierIsForBindingPolicy(objIdentifier) {
+		if err := c.handleBindingPolicy(ctx, objIdentifier); err != nil {
 			return fmt.Errorf("failed to handle bindingpolicy: %w", err) // error logging after this call
 			// will add name.
 		}
 
-		logger.Info("Handled bindingpolicy", "object", util.RefToRuntimeObj(obj))
+		logger.Info("Handled bindingpolicy", "object identifier", objIdentifier)
 		return nil
-	} else if util.IsCRD(obj) {
-		if err := c.handleCRD(obj); err != nil {
+	} else if util.ObjIdentifierIsForCRD(objIdentifier) {
+		if err := c.handleCRD(ctx, objIdentifier); err != nil {
 			return fmt.Errorf("failed to handle CRD: %w", err) // error logging after this call
 			// will add name.
 		}
-		logger.Info("Handled CRD", "object", util.RefToRuntimeObj(obj))
+		logger.Info("Handled CRD", "object identifier", objIdentifier)
 	}
 
-	// avoid further processing for keys of objects being deleted that do not have a deleted object
-	if isBeingDeleted(obj) && key.DeletedObject == nil {
-		return nil
-	}
-
-	return c.updateResolutions(ctx, obj)
+	return c.updateResolutions(ctx, objIdentifier)
 }
 
-func (c *Controller) getObjectFromKey(key util.Key) (runtime.Object, error) {
-	lister := c.listers[key.GvkKey()]
+func (c *Controller) getObjectFromObjectIdentifier(objIdentifier util.ObjectIdentifier) (runtime.Object, error) {
```

Comment:
FYI: I think the name of this func needs to say "object" twice; I would call it "getObjectFromIdentifier" or maybe even just "getObject".